### ISSUE TITLE
Association definitions

### ISSUE THREAD

**Issue by cheerfulstoic** [neo4jrb/neo4j#998]
Often when I'm writing associations I think this and so I just need to get it off my chest:

I always stumble when writing associations because I have to define a direction before I define a type, but I don't know the direction until I've decided on a type.

Anybody else have this problem?

**Comment by bloomdido** [comment]
`has_one` and `has_many` really screw me up, as I expect the name of the association to come first, just as in Rails. Seems to me that this syntax would be much preferable:

```
has_many :members, type: "MEMBER_OF", dir: :in, model_class: :any
```

By default, the `dir` could be `:both` if it is not specified. Also, I always screw up the `model_class` param. It seems like `:any` should be used instead of `false`, like it is in ActiveRel. It should also be the default for `type` and `model_class` when not specified. So:

```
has_many :members
```

would functionally be the same as:

```
has_many :both, :members, type: false, model_class: false
```

This would allow you to add `dir`, `type`, and `model_class` as needed.

Just my $0.02.

**Comment by cheerfulstoic** [comment]
We actually starting off in the gem with having the `dir` as an option.  But `:both` is actually a pretty rare case, so you actually are specifying the direction almost every time.

I definitely think that having the association name first, though.  

I think that you can use either `false` or `:any` (if not, I think there might be a bug).  Again, though, in my experience specifying a relationship type / model class are what is done most of the time, not as the default.  Being able to not type as much in some cases is nice, but if it confuses people who are just starting that's a big disadvantage.

That said, I'm **very** sympathetic to shorter associations and more convention-over-configuration (where it makes sense).  Even though I'm the maintainer, it still feels somewhat more awkward compared to `ActiveRecord`.

Let me know if this doesn't match with your experience or if you have other ideas / thoughts.

**Comment by bloomdido** [comment]
I can definitely see the reasoning behind forcing all the params to declared in the association. I don't think it's wrong - just one opinionated approach. My instincts lean toward declaring less options should result in the widest result set. While an association with `:both` as the direction might not be what you want most of the time, it does work. Same with not declaring the `:type` or `:model_class`. In fact, if I create the following association, I will get back all connected nodes:

```
has_many :both, :linked_nodes, type: false, model_class: false
```

Starting with that, I feel like I can start narrowing that down my results by adding more specific conditions. That's why, to me, something like `has_many :linked_nodes` feels like an appropriate starting place. It's just my bias however and certainly not any more or less valid than anyone else's.

**Comment by bloomdido** [comment]
Regarding `type: :any` and `model_class: :any` - I don't think that will work, as both conditions accept symbols that are mapped to strings/classes. It would create a conflict with a relationship type "ANY" or class named "Any". It just feels awkward because ActiveRel uses `:any` in its declarations. Not sure anything can be done about it now.

**Comment by cheerfulstoic** [comment]
For `model_class: :any` I don't think it's a problem because `any` can't be a Ruby constant (the first letter isn't capitalized).  For relationship types, there could be something there.  Though I think if you used the string `any` it would work, but it might not be ideal.

I definitely get what you're saying about being able to only specify what is needed.  I could certainly see things being fine for new users if almost all parts the documentation showed a `type` / `model_class` declared.  I do worry, though, because it was definitely confusing for people when we had it that way originally, but there might be a good way to do it.

Definitely interested in feedback from @subvertallchris and @ProGM on this one

**Comment by subvertallchris** [comment]
I think that the extra configuration required for each association provides models that are easier to read and data models that always match the user's expectations, even if it means a little extra configuration upfront. The disconnect between what the user expects the default to be and what it actually is hurts new users particularly hard. Once you recognize your mistake, the presence of the opinionated default requires just a little extra thought with all future calls to those methods to make sure that you _actually_ override that option so you end up with what you really want.

**Comment by cheerfulstoic** [comment]
When you say it makes it easier to scan, I'm assuming that you're not scanning the directions because those are just a detail of the relationship type you chose.  The easy scanning comes from being able to read the association names becaues they are lined up, right?  Though they aren't exactly because `in`, `out`, and `both` are different numbers of characters.

I'm definitely somewhere in the middle.  I miss the simplicity of AR associations.  I even still hope for some solution, in my heart-of-hearts, where we could have default relationship types based off the model involved so that people could start using the gem without needing to think about the relationship types too much.  Yes, understanding Neo4j details is important, but it could ease the startup.  Also, now that @ProGM has made a nice migration system, it would be easy to change relationship types if you decided you wanted something else.

An example of what I'm thinking of:

```ruby
class Post
  has_many :comments # defaults to direction `out` and type `HAS_COMMENT`
  has_one :comments # also defaults to direction `out` and type `HAS_COMMENT`

  belongs_to_many :comments # defaults to direction `in` and type `HAS_POST`
  belongs_to :comments # defaults to direction `in` and type `HAS_POST`
end
```

This would line up very nicely with the AR idea of table name / column name conventions.

**Comment by subvertallchris** [comment]
Woah hold on, they have different numbers of characters? That changes things. ;-) You don't think it still keeps things tidier than leading with association, which would potentially require a lot of extra whitespace to keep things lined up and quickly readable?

**Comment by bloomdido** [comment]
Well, I think that just proves that it's going to be very difficult to get any "defaults" right. I am very much not a fan of "HAS_XXX" named links. I much prefer naming my links such that they "read" in the direction of the link.

i.e.
```
class User
  has_many :groups, type: "MEMBER_OF", dir: :out
end

class Group
  has_many :members, type: "MEMBER_OF", dir: :in
end
```

It took me a while to favor this style of naming, but until I did, I was constantly confused as to which direction a link should point and what its name should be. I feel as though the approach @cheerfulstoic  has recommend above will further propagate that confusion for newbies.

I suppose that's why I'm inclined to believe the simplest version should be agnostic. `has_many :members` should name the relationship `MEMBER` and not pick a direction. I agree, it won't help the user choose a style, but it also won't adopt one for them. In this situation (and mapping data in a graph db in general), I feel like there is no "correct" choice. Hence, the idea of a default picking one annoys me because if I choose a style that opposes the default, I feel like I'm always fighting the system. Again, it's a pretty minor beef, but if I feel like the defaults are "wrong", I'm never going to quite feel comfortable.

I do, however, feel very strongly that the name of the association should come first. After I have defined the relationship, I no longer care what direction it points or the name of type - I only care about the name of the association because that's what I'm going to be using in my code. I will be writing `user.groups` or `group.members`, and so that's what I want to be able to quickly see.

**Comment by bloomdido** [comment]
Also, 👎 👎 👎 for `belongs_to`. It's one of the things that's driven me most crazy with AR. I've been programming with Rails for 10 years and I STILL have to pause every time I create a relationship to figure out if its the `has_xxx` or the `belongs_to` side. With a SQL DB, it's necessary because you need to know which side has the primary key and which side has the foreign key. But with Neo4j, there's no difference in the nodes themselves, as the relationship spells that out, and so you're adding confusion for no reason whatsoever.
/opinion_off ;-)

**Comment by bloomdido** [comment]
Come to think about it, I'm not sure I think there should be `has_one` and `has_many`, because with Neo4j, there's no difference. It should just be `has`. The only difference is how ActiveNode caches the associated nodes, and whether the accessors accept/return an array or single node. Of course, I hide all my associations behind an underscore now (`has_many :_members`) because it's the best way I've found of overriding accessors (though that's hopefully going to change soon when I can get around to updating the behavior of `read_attribute` and `write_attribute`.
/back_to_work

**Comment by cheerfulstoic** [comment]
Agreed about the association name being first.

Also agreed that `belongs_to` is confusing with AR and I also often have to think twice about it, *but* we don't have foreign keys in Neo4j, so it's much simpler to think of "`belongs_to` means the direction points to to this model and `has` points away".

@subvertallchris maybe you can help me remember why we moved towards a required `type`.  IIRC, the thought was that we're not going to be able to come up with a good relationship type by default.  I definitely agree that the best relationship type is something that makes a sentence and `(:Person)-[:MEMBER]-(:Group)` isn't as good as `MEMBER_OF` and we want to encourage good relationship naming practices in Neo4j.  It would be out of our scope, I think, to figure out how to find a good english-language relationship.

And you are right, that Neo4j doesn't draw a distinction between `has_one` and `has_many`, but it's been incredibly valuable to do that in the gem.  But I definitely would also like a way to start an association chain on a has_one association.  Something perhaps like a `member_proxy` if you've defined a `has_one` for `member`

**Comment by bloomdido** [comment]
I agree that `has_many` and `has_one` are useful in ActiveNode's context. I was just pontificating on reasons for not trying to follow ActiveRecord too closely. 😄

Since I personally define all of my relationships using ActiveRel, I don't really use the `type` attribute. But without doing something like what `ActiveSupport::Inflector` does for plural and singular, I'm not sure you'll ever really be able to come up with a well-named auto type. And even then, there are difficult ones even for a person to figure out. For example, I have a relationship that maps an event to a calendar tree. If the event is mapped to a Day node, the appropriate name is "OCCURS_ON", but if it's a time, "OCCURS_AT" is better. I thought about going with "OCCURS_XX", but that seemed just too awkward, so I just went with "OCCURS", which works because the sentence only makes sense in one direction: the event OCCURS on a date, but a date cannot occur on an event. Something to do with the relationship name being a verb or something, though it doesn't always work that way. 😆

**Comment by subvertallchris** [comment]
We moved to a required type because the gem's provided default wasn't adequate. We went from lowercase association names to all caps but users still found it surprising. Maybe we just hadn't documented it clearly enough.
